Improve conditionalExpressionTypes

rajyan opened the pull request:
Steps left
- [x] permit all expression types as conditional expression keys
- [x] change the type guard logic by using the specified scope
- [ ] Improve `createConditionalExpressions`
- [x] fix conditional expression invalidations
- [x] use `ExpressionTypeHolders` for conditions
- [ ] change `addConditionalExpressions` first arg to expr


fixes https://github.com/phpstan/phpstan/issues/3677
fixes https://github.com/phpstan/phpstan/issues/5623
fixes https://github.com/phpstan/phpstan/issues/5401 (the playground in the comment is another issue, we should open a new issue)
fixes https://github.com/phpstan/phpstan/issues/7292
fixes https://github.com/phpstan/phpstan/issues/8212

rajyan committed:
add foreach conditional expression for ArrayDimFetch

rajyan commented:
Oh, failed rebasing.., :cry:

rajyan committed:
conditional expression types for all expressions

rajyan committed:
use new scope type to compare instead of type guards

rajyan committed:
fix tests because rand is not marked impure

rajyan committed:
preserve the original test by adding @phpstan-impure

rajyan commented:
Sorry miss clicked

rajyan commented on src/Analyser/MutatingScope.php:
```diff
@@ -3758,51 +3721,20 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 					}
-
-					if (!$typeGuards[$conditionExprString]->equals($conditionalType)) {
+					if (!$scope->getType($conditionExpr)->equals($conditionalType)) {
```
This line is causing a infinite recursion, not sure why yet.

rajyan replied:
recursion with this line
https://github.com/rajyan/phpstan-src/blob/0ae0fe306679e400843bcf2c6de36f67be1eaa9b/src/Analyser/MutatingScope.php#L1601-L1603

rajyan replied:
Solved

rajyan committed:
filterByFalseyValue only if issetCheck is falsey

rajyan commented:
I think we can clean up the conditional expression logic, and make it easier to migrate to LateResolvableTypes

rajyan commented:
Memo
https://github.com/phpstan/phpstan-src/blob/579b075806f32076dfaf55a4f2c913f55ed14a5c/tests/PHPStan/Analyser/data/bug-3677.php#L75

https://github.com/rajyan/phpstan-src/blob/e4912564467d603ad64fff988c1249fd6d059fa6/tests/PHPStan/Analyser/data/dependent-variable-certainty.php#L261-L480

rajyan commented:
Steps left
1. Improve `createConditionalExpressions`
2. fix conditional expression invalidations
3. use `ExpressionTypeHolders` for conditions
4. change `addConditionalExpressions` first arg to expr

rajyan commented:
Another note,

We need to carefully separate assign and specifying
```php
// when $b is conditional of $a = 'foo' => 'foo', $a = 'bar' => 'bar'
echo $b; // 'foo'|'bar'
if ($a === 'foo') {
  echo $b; // 'foo'
}
if ($a === 'bar') {
  echo $b; // 'bar'
}
if ($a === 'foo') {
  $a = 'bar';
  echo $b; // 'foo'
}
$a = 'foo';
echo $b;  // 'foo'|'bar'
```

Assign is a overwrite and needs invalidation of related resolved expression types, while specifying should be an intersection and no need to invalidate in my opinion.

rajyan committed:
no need to filter conditional types when specifying

rajyan committed:
temporary fix

rajyan committed:
fix improved test result

ondrejmirtes commented:
BTW just a minor nitpick - "temporal" is related to time, "temporary" is what you want here :)

ondrejmirtes commented:
BTW there are some commented-out tests here https://github.com/phpstan/phpstan-src/blob/1.9.x/tests/PHPStan/Analyser/data/dependent-variable-certainty.php, maybe you could debug that as part of this or some future PR :)

ondrejmirtes commented:
This fixes some issues: https://github.com/phpstan/phpstan-src/actions/runs/3420191417

Would be nice to add regression tests here for those which are fixed, to make sure they don't get broken in future iterations.

rajyan commented:
@ondrejmirtes 
The new issue bot CI looks so wonderful!👍 
Thank you for it.

rajyan committed:
add regression tests detected by issue bot

rajyan commented:
Memo
SpecifyExpressionType and assignExpression with certanity maybe

rajyan committed:
separate assigning and specifying correctly
only invalidate conditionalExpressions in assigning

ondrejmirtes commented:
The issue bot looks like there are some performance degradations but if I try it locally it's all fast 🤷 

<details>

```
Analysis of ee1dace3-b155-46ca-a2d3-23b2a164f1fe took 1.23 s
Analysis of 4f48d363-0ad9-4f78-837c-049f290adfa1 took 1.32 s
Analysis of 9b4266c2-5fc6-4327-ae59-e6aa376a53bc took 1.32 s
Analysis of d548d784-9771-42c8-95fc-66e8cf165abd took 1.22 s
Analysis of 0f73146e-e143-47c0-9e3a-5463e375ea71 took 1.04 s
Analysis of 403a5ffa-260f-4bfa-b8f0-4959bb07e917 took 1.04 s
Analysis of b22de8cd-024a-475b-8ea7-6601df32b245 took 1.38 s
Analysis of 6a9d7024-7818-4de5-b224-961eebcda927 took 1.02 s
Analysis of 88facd1a-c1f6-4bf6-bae2-cc13dd1037f5 took 1.04 s
Analysis of 32d2c030-666b-4e37-94bf-2f82ca9ce28f took 1.30 s
Analysis of d5591f7c-b4e1-40bc-adce-15e4d846a4f7 took 1.27 s
Analysis of a5df0bc8-e075-405c-badb-f378151c88b8 took 1.33 s
Analysis of 4f814a09-c46e-469d-aea7-7c6491a8547e took 1.33 s
Analysis of 213421a3-bdf9-4ee1-93b9-6e3cc368021a took 1.38 s
Analysis of 59007ebf-a020-4e0d-884e-0e5854a3b925 took 1.49 s
Analysis of ab2f54a7-21eb-4602-b4c6-06f6c618896a took 1.26 s
Analysis of 84d06ae8-437b-4b90-9985-7a236da592cd took 1.23 s
Analysis of ced49821-322b-4766-93c1-7fb720b453d4 took 1.21 s
Analysis of 72de1121-4c43-4d6b-9565-cbd73456d2d3 took 1.21 s
Analysis of 64ff1d9b-e885-4987-91f9-6dd71b436e74 took 1.20 s
Analysis of 3ffce6d8-8337-4500-860d-04def4625ae5 took 1.21 s
Analysis of 5b8be542-c44c-40df-b66a-99a8e5cb0d03 took 1.25 s
Analysis of fc89833b-44e1-4f98-9651-cb2f8c4a887c took 1.32 s
Analysis of d426c53a-8ada-4f53-a719-3ca296c3459d took 1.21 s
Analysis of 348c3626-bfe6-4f25-9085-3684be966060 took 1.20 s
Analysis of 283e2e30-07b8-4811-9dc5-f574237291d3 took 1.22 s
Analysis of c94175a1-1278-4f3d-9424-fe069e1bc5ec took 1.21 s
Analysis of ab0c3d35-e645-449c-b27c-fbc089b784b2 took 1.21 s
Analysis of 5003d274-3690-464e-a5e1-6e898cd00411 took 1.22 s
Analysis of 892c0769-7920-42cf-9354-69c91b915ef1 took 1.22 s
Analysis of 4cc8f7c3-c43f-478d-ad29-dfdb1f168c04 took 1.21 s
Analysis of 97f95150-498a-4722-a0fb-cbe13af32bf3 took 1.37 s
Analysis of 24aaa27a-e6a5-478b-a85f-b25af078d5af took 1.37 s
Analysis of adeaffc9-e53c-48ee-93aa-748069f6308b took 1.24 s
Analysis of f5f77c8d-2329-4ec4-ac9c-4fde412fd663 took 1.24 s
Analysis of 0c1858ce-f4bc-45cc-8742-3a161b0f0022 took 1.35 s
Analysis of 2c774269-8a82-40b0-af4f-67b1b8318857 took 1.34 s
Analysis of 44da1ed6-dcaf-4139-a747-1ac568c71f14 took 1.34 s
Analysis of c772c666-f8d8-4676-b1c4-8d6d3bfee327 took 1.27 s
Analysis of 1570736d-c072-45a2-bad5-51ed6f41454a took 1.36 s
Analysis of 6967bdde-8c8f-476d-adcd-c94db723cd8b took 1.26 s
Analysis of d0a6e763-62c0-42f2-a99f-6817809609bd took 1.23 s
Analysis of 3a219790-5e24-4d7e-b245-cc9e9dbc510d took 1.36 s
Analysis of 32148a16-d0c2-4b1b-9e55-3f784ea8d81f took 1.37 s
Analysis of ed660810-bbd7-492e-b430-b47f79f35f55 took 1.36 s
Analysis of 4ec591fe-7fa0-4d79-b8ae-e3df5eadb92e took 1.21 s
Analysis of 2c2e6e2a-9568-49f8-813b-e4c89eac95a9 took 1.22 s
Analysis of 9ed5bd67-660e-4084-9848-13db4201e2c5 took 1.21 s
Analysis of 0d40f950-1615-4673-b032-dbdd19dfdf2d took 1.21 s
Analysis of 073a2f00-7331-4a34-8bcc-78dc39f0319f took 1.24
```

</details>

rajyan commented:
Not sure enough what’s happening, but maybe something is wrong with the commented out tests. I added them because the issue bot said it was fixed, but it wasn’t. https://github.com/phpstan/phpstan-src/pull/1950/commits/4e753368118d17bfb5e7400c49624b056d921a80

rajyan commented:
I think testBug7705 is causing an infinite loop and that may be the cause.

rajyan commented on src/Analyser/MutatingScope.php:
```diff
@@ -3756,128 +3725,79 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
-	 * @param array<string, ConditionalExpressionHolder[]> $newConditionalExpressionHolders
-	 */
-	public function changeConditionalExpressions(array $newConditionalExpressionHolders): self
+	private function exprStringToExpr(string $exprString): ?Expr
```
This is added temporary. Could be deleted after ConditionalExpressionHolder condition types is changed to ExpressionTypeHolders

ondrejmirtes replied:
Yeah, I'd welcome that :)

rajyan commented:
Todos
* solve conflict with https://github.com/phpstan/phpstan-src/commit/582a9cb8b9b4fce2bd069bac26bf1d31dd52e7e2
* https://github.com/phpstan/phpstan-src/pull/1950#discussion_r1019755872
* solve infinite loop? in https://github.com/phpstan/phpstan-src/pull/1950/commits/5600e4930b59ff6f5f8a752837f3c837452e4595 comment out tests

rajyan commented:
The infinite loop is caused by `filterBySpecifiedTypes` of `BinaryOp` => `getType` of conditional expression of `BinaryOp` => `getType` of `BinaryOp` uses `filterBySpecifiedTypes` ...
![image](https://user-images.githubusercontent.com/38206553/201503360-d638e6d6-1429-483a-86fd-df48410a01c0.png)

rajyan committed:
Revert "fix tests because rand is not marked impure"

This reverts commit 132b39445de42829ff8dcfd07df34e634dcf8fc6.

rajyan committed:
Revert "preserve the original test by adding @phpstan-impure"

This reverts commit ee0bbfac7d75ee79de32e19b2dbcb8e8714b1960.

rajyan committed:
revert changes in createConditionalExpressions

rajyan committed:
add regression tests from issuebot

rajyan commented:
I'm gonna revert changes in `createConditionalExpressions` of `mergeWith` for now, because the pull request is getting to large... The changes in this PR can fix a lot of dependant type issues already:smile:

rajyan commented:
What I did in this pull request
1. Separated specifying and assigning
  * Only invalidate conditionalExpressions in assigning
2. Changed logic of filtering conditionalExpressions
  * Use getType when to check matching expressions (to make conditionalExpressions work for all expressions)
3. Refactored the unsetExpression logic

rajyan commented:
I think this pull request is mergable now.
The steps left are
- [x]  Use `ExpressionTypeHolders` for `ConditionalExpressionTypes` conditions (to get rid of `exprStringToExpr`)
- [ ] Make `createConditionalExpressions` to work for all expressions
- [ ] Make `processBooleanConditionalTypes` to work for all expressions
- [ ] improve `addConditionalExpressions`

and then, we can change `ConditionalExpressions` to `LateResolvableType`.

rajyan commented:
The phpunit integration test failure seems related

rajyan commented:
Found that specified types from `getNewConditionalExpressionHolders` is wrong in some cases.
I'll create another pull request and fix it there.

rajyan commented:
https://github.com/phpstan/phpstan-src/blob/0936b3b5b739cf990dd4d386ae5e73a4225afa8f/src/Analyser/TypeSpecifier.php#L746-L749
This line is wrong. It's creating the same conditional expression types as && from ||.

rajyan commented:
What I'm saying wrong, is this
![image](https://user-images.githubusercontent.com/38206553/201517821-8ca2f1c3-6f43-4737-bab8-f89de8dbb952.png)
created from
```php
if ($loaded === false || ($strict && $message !== '')) {
```

rajyan committed:
fix and improve subtracting types

rajyan commented:
Fixed another bug and I think this is finally ready!

rajyan commented:
The error of rector is legit, but the error message should be improved.
https://github.com/rectorphp/rector-src/blob/71b1142ee41eceef21d9c3d3a1bf5bc4f67ad9d2/packages/NodeTypeResolver/PHPStan/TypeHasher.php#L122
```php
! $currentType instanceof AliasedObjectType
```
is always false, because it's narrowed in the previous line.

rajyan committed:
remove comments now fixed

herndlm commented:
Does it make sense to also add `assertType()` checks to the dependable type related fixed issues? I was looking forward to those in the diff :) The same issue files could be reused in NodeScopeResolverTest.

ondrejmirtes commented:
Awesome, thank you! Looking to other improvements :)

ondrejmirtes commented:
What doesn't sit well with me and what you haven't changed yet is the places in NodeScopeResolver where `addConditionalExpressions` are called. It's only for variables so it doesn't work generally for all expressions :)

rajyan commented:
@herndlm 
Yeah, I'll add them!

@ondrejmirtes 
Created a ConditionalExpressionTypes todos to remember what to do, and make others can work on it too if someone is interested. 
https://github.com/phpstan/phpstan/issues/8360
